src/github.com/couchbase/sync_gateway/db/index_changes.go

```diff
@@ -286,13 +229,123 @@ func (db *Database) VectorMultiChangesFeed(chans base.Set, options ChangesOption
 				output <- &change
 				return
 			}
-			writeHistogram(indexTimingExpvars, iterationStartTime, "index_changes_iterationTime")
 		}
 	}()
 
 	return output, nil
 }
 
+// Determines whether the clock hash should be calculated for the entry. For non-continuous changes feeds, hash is only calculated for
+// the last entry sent (for use in last_seq), and is done in the defer for the main VectorMultiChangesFeed.
+// For continuous changes feeds, we want to calculate the hash for every nth entry, where n=kContinuousHashFrequency.  To ensure that
+// clients can restart a new changes feed based on any sequence in the continuous feed, we set the last hash calculated as the LowHash
+// value on the sequence.
+func (db *Database) calculateHashWhenNeeded(options ChangesOptions, entry *ChangeEntry, cumulativeClock base.SequenceClock, hashedEntryCount *int, continuousLastHash string) string {
+
+	if options.Continuous != true {
+		// For non-continuous, only need to calculate hash for lastSent entry.  Initialize empty clock and return
+		entry.Seq.Clock = base.NewSyncSequenceClock()
+	} else {
+		// When hashedEntryCount == 0, recalculate hash
+		if *hashedEntryCount == 0 {
+			clockHash, err := db.SequenceHasher.GetHash(cumulativeClock)
+			if err != nil {
+				base.Warn("Error calculating hash for clock:%v", base.PrintClock(cumulativeClock))
+				return continuousLastHash
+			} else {
+				entry.Seq.Clock = base.NewSyncSequenceClock()
+				entry.Seq.Clock.SetHashedValue(clockHash)
+				continuousLastHash = clockHash
+			}
+			*hashedEntryCount = kContinuousHashFrequency
+		} else {
+			entry.Seq.LowHash = continuousLastHash
+			*hashedEntryCount--
+		}
+	}
+	return continuousLastHash
+}
+
+// Creates a go-channel of ChangeEntry for each channel in channelsSince.  Each go-channel sends the ordered entries for that channel.
+func (db *Database) initializeChannelFeeds(channelsSince channels.TimedSet, options ChangesOptions, addedChannels base.Set, userVbNo uint16) ([]<-chan *ChangeEntry, error) {
+	// Populate the  array of feed channels:
+	feeds := make([]<-chan *ChangeEntry, 0, len(channelsSince))
+
+	base.LogTo("Changes+", "GotChannelSince... %v", channelsSince)
+	for name, vbSeqAddedAt := range channelsSince {
+		seqAddedAt := vbSeqAddedAt.Sequence
+		// If there's no vbNo on the channelsSince, it indicates a user doc channel grant - use the userVbNo.
+		var vbAddedAt uint16
```

Comment:
does `vbAddedAt` contain a vbucket number or a per-vbucket sequence?  the "AddedAt" sounds like a point of time, which would indicate it contained a per-vbucket sequence.  

I realize this is probably too big of a change for this PR, but it might be nice if we added some named type aliases to easily identify what the expected contents of variables are .. eg:

```
type vbucketId uint16
type vbucketSeq uint16
+func (db *Database) initializeChannelFeeds(channelsSince channels.TimedSet, options ChangesOptions, addedChannels base.Set, userVbNo vbucketId) ([]<-chan *ChangeEntry, error) {
   ...
   var vbAddedAt vbucketSeq
}
```

that would also make it easy